File: sdks/typescript/src/auth-uuid.ts
import type { CreateDeckOptions } from './types.js';

export const AUTH_UUID_STORAGE_KEY = 'df_uuid';
export const AUTH_UUID_FILENAME = 'auth-uuid';

const UUID_V4_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;

type BrowserStorage = {
  getItem(key: string): string | null;
  setItem(key: string, value: string): void;
};

export function isValidAuthUuid(value: string | null | undefined): value is string {
  return typeof value === 'string' && UUID_V4_RE.test(value);
}

export function generateAuthUuid(): string {
  return globalThis.crypto.randomUUID();
}

function isNode(): boolean {
  return typeof process !== 'undefined' && process.versions != null && process.versions.node != null;
}

function isBrowserWithLocalStorage(): boolean {
  try {
    return getBrowserStorage() !== undefined;
  } catch {
    return false;
  }
}

function getBrowserStorage(): BrowserStorage | undefined {
  return (globalThis as typeof globalThis & { localStorage?: BrowserStorage }).localStorage;
}

let memoryFallback: string | undefined;
let defaultResolving: Promise<string> | undefined;

function readBrowserStorage(): string | null {
  if (!isBrowserWithLocalStorage()) {
    return null;
  }
  try {
    const value = getBrowserStorage()?.getItem(AUTH_UUID_STORAGE_KEY) ?? null;
    return isValidAuthUuid(value) ? value : null;
  } catch {
    return null;
  }
}

function writeBrowserStorage(value: string): void {
  if (!isBrowserWithLocalStorage()) {
    return;
  }
  try {
    getBrowserStorage()?.setItem(AUTH_UUID_STORAGE_KEY, value);
  } catch {
    // Storage may be unavailable in private browsing mode.
  }
}

async function getNodeConfigDir(): Promise<string> {
  const os = await import('node:os');
  const path = await import('node:path');
  return (
    process.env.DECKFLOW_CONFIG_DIR ||
    path.join(os.homedir(), '.deckflow')
  );
}

async function readNodeFile(): Promise<string | null> {
  if (!isNode()) {
    return null;
  }
  try {
    const fs = await import('node:fs/promises');
    const path = await import('node:path');
    const configDir = await getNodeConfigDir();
    const filePath = path.join(configDir, AUTH_UUID_FILENAME);
    const content = (await fs.readFile(filePath, 'utf-8')).trim();
    return isValidAuthUuid(content) ? content : null;
  } catch {
    return null;
  }
}

async function writeNodeFile(value: string): Promise<void> {
  if (!isNode()) {
    return;
  }
  try {
    const fs = await import('node:fs/promises');
    const path = await import('node:path');
    const configDir = await getNodeConfigDir();
    const filePath = path.join(configDir, AUTH_UUID_FILENAME);
    await fs.mkdir(configDir, { recursive: true });
    await fs.writeFile(filePath, value, 'utf-8');
  } catch {
    // File system may be read-only.
  }
}

async function readFromDefaultStorage(): Promise<string | null> {
  if (isBrowserWithLocalStorage()) {
    return readBrowserStorage();
  }
  if (isNode()) {
    return await readNodeFile();
  }
  return null;
}

async function persistToDefaultStorage(value: string): Promise<void> {
  if (isBrowserWithLocalStorage()) {
    writeBrowserStorage(value);
    return;
  }
  if (isNode()) {
    await writeNodeFile(value);
  }
}

async function resolveWithCustomStorage(options: CreateDeckOptions): Promise<string> {
  const storage = options.authUuidStorage;
  if (!storage) {
    throw new Error('authUuidStorage is required');
  }

  const stored = await storage.get();
  if (isValidAuthUuid(stored)) {
    return stored;
  }

  const generated = generateAuthUuid();
  await storage.set(generated);
  return generated;
}

async function resolveWithDefaultStorage(): Promise<string> {
  const stored = await readFromDefaultStorage();
  if (stored) {
    return stored;
  }

  const generated = generateAuthUuid();
  await persistToDefaultStorage(generated);
  memoryFallback = memoryFallback ?? generated;
  return generated;
}

export async function resolveAuthUuid(options: CreateDeckOptions = {}): Promise<string> {
  if (isValidAuthUuid(options.authUuid)) {
    return options.authUuid;
  }

  if (isNode() && isValidAuthUuid(process.env.DECKTOOLS_AUTH_UUID)) {
    return process.env.DECKTOOLS_AUTH_UUID;
  }

  if (options.authUuidStorage) {
    return resolveWithCustomStorage(options);
  }

  if (!defaultResolving) {
    defaultResolving = resolveWithDefaultStorage();
  }
  return defaultResolving;
}

/** Resets module-level caches. For unit tests only. */
export function resetAuthUuidCacheForTests(): void {
  defaultResolving = undefined;
  memoryFallback = undefined;
}

File: sdks/go/auth_uuid.go
package decktools

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
)

const (
	AuthUUIDStorageKey = "df_uuid"
	AuthUUIDFilename   = "auth-uuid"
)

var (
	uuidV4RE       = regexp.MustCompile(`(?i)^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`)
	defaultUUIDMu  sync.Mutex
	defaultUUIDVal string
)

func IsValidAuthUUID(value string) bool {
	return uuidV4RE.MatchString(value)
}

func GenerateAuthUUID() (string, error) {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		return "", err
	}
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80

	dst := make([]byte, 36)
	hex.Encode(dst[0:8], b[0:4])
	dst[8] = '-'
	hex.Encode(dst[9:13], b[4:6])
	dst[13] = '-'
	hex.Encode(dst[14:18], b[6:8])
	dst[18] = '-'
	hex.Encode(dst[19:23], b[8:10])
	dst[23] = '-'
	hex.Encode(dst[24:36], b[10:16])
	return string(dst), nil
}

func resolveAuthUUID(ctx context.Context, options ClientOptions) (string, error) {
	if IsValidAuthUUID(options.AuthUUID) {
		return options.AuthUUID, nil
	}
	if env := os.Getenv("DECKTOOLS_AUTH_UUID"); IsValidAuthUUID(env) {
		return env, nil
	}
	if options.AuthUUIDStorage != nil {
		return resolveAuthUUIDWithStorage(ctx, options.AuthUUIDStorage)
	}

	defaultUUIDMu.Lock()
	defer defaultUUIDMu.Unlock()
	if IsValidAuthUUID(defaultUUIDVal) {
		return defaultUUIDVal, nil
	}

	if stored, _ := readDefaultAuthUUID(); IsValidAuthUUID(stored) {
		defaultUUIDVal = stored
		return stored, nil
	}

	generated, err := GenerateAuthUUID()
	if err != nil {
		return "", err
	}
	_ = writeDefaultAuthUUID(generated)
	defaultUUIDVal = generated
	return generated, nil
}

func resolveAuthUUIDWithStorage(ctx context.Context, storage AuthUUIDStorage) (string, error) {
	stored, err := storage.Get(ctx)
	if err != nil {
		return "", err
	}
	if IsValidAuthUUID(stored) {
		return stored, nil
	}

	generated, err := GenerateAuthUUID()
	if err != nil {
		return "", err
	}
	if err := storage.Set(ctx, generated); err != nil {
		return "", err
	}
	return generated, nil
}

func defaultConfigDir() string {
	for _, key := range []string{"DECKFLOW_CONFIG_DIR"} {
		if dir := os.Getenv(key); dir != "" {
			return dir
		}
	}
	home, err := os.UserHomeDir()
	if err != nil || home == "" {
		return ""
	}
	return filepath.Join(home, ".deckflow")
}

func readDefaultAuthUUID() (string, error) {
	dir := defaultConfigDir()
	if dir == "" {
		return "", nil
	}
	content, err := os.ReadFile(filepath.Join(dir, AuthUUIDFilename))
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(content)), nil
}

func writeDefaultAuthUUID(value string) error {
	dir := defaultConfigDir()
	if dir == "" {
		return nil
	}
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(dir, AuthUUIDFilename), []byte(value), 0o600)
}

func resetAuthUUIDCacheForTests() {
	defaultUUIDMu.Lock()
	defer defaultUUIDMu.Unlock()
	defaultUUIDVal = ""
}
